### ISSUE TITLE
Automatic purging of old releases, based on PAUSE user config

### ISSUE THREAD

**Issue by neilb** [andk/pause#223]
I know this has been discussed a number of times, and is a hairy topic, but I'm opening this here as a place to capture thoughts, and a reminder to discuss it at next year's QAH, if not sooner.

I imagine this requires

 * thinking about the different options that need to be supported and the rules related to them
 * added a new field to PAUSE users, where they can select the purge option, defaulting to "none".
 * implementing the purge code including sending of notifications
 * once this is done, emailing all authors encouraging them to enable auto purge

The rules / options will be the interesting discussion. For example, my general approach is:

 * Keep the most recent developer release, if it's more recent that the most recent non-developer release
 * Keep the most recent non-developer release
 * Once a developer release has been superseded, and a week or so has passed, then ok to delete
    older releases.

I know people use different rules, for example keeping the last N releases, and that there are sometimes special cases, for example where you want to keep the last stable version of a 1.X release on CPAN, even if it was long-since superseded by 2.X releases. Though some people address that by renaming the dist, eg `Dancer` and `Dancer2`.

**Comment by xsawyerx** [comment]
Other than `perl`, I don't think you can provide two different "production" versions. That's one of the considerations for `Dancer` vs. `Dancer2`. We could not release both Dancer 1.x and Dancer 2.x. PAUSE would have problems with the 1.X version. `perl` is a special case.

I think deciding on the different options users can have and then adding buttons that allow them to be included in a cron script. The cron script runs once a day/week/month and decides what to remove. 

While sending alerts every once and a while when it gets filled up, I think it would be a better long-term to allow users to say "Yes, it's fine to delete for me. Just leave me with the last N stable ones." and get it over with.

**Comment by ribasushi** [comment]
> Other than perl, I don't think you can provide two different "production" versions.

The user PAUSE directory is nothing more than a "dropbox": one can ([and does](http://lists.scsys.co.uk/pipermail/dbix-class/2016-June/012168.html)) upload older versions (`cpanm --look DBIx::Class'~<0.0828'` works as expected)

**Comment by rjbs** [comment]
What problem is this meant to solve?

**Comment by rjbs** [comment]
I was asked to clarify the above.

What is the point of auto-cleaning author directories?  The CPAN admins plan to increase storage.  What's our rate of growth, and its derivative?  Is it such that we're still likely to need to undertake this project once the available storage is increased?

I worry that this is going to be a bunch of work and policy creation that isn't really needed.

**Comment by ribasushi** [comment]
> I worry that this is going to be a bunch of work and policy creation that isn't really needed.

+1, I feel almost entirely the same. The **smalest** SSD one can realistically purchase today (left alone in couple years) is larger than 100GB.

The entire master archive currently weighs way under that, with no signs of impending "hockey stick".

**Comment by neilb** [comment]
I can think of at least two benefits:

 * less diskspace (as you say, this probably isn't compelling enough)
 * when people refactor dists and change module names, we end up with the old releases still being indexed, with the potential for the wrong release being installed. Ok, the probability is low, but auto cleaning up would make it even lower.

In working through the permissions conflict, I've hit a number of these. A quick script tells me there are just over 1500 cases in the current index.

As @rjbs says, maybe the work still isn't warranted, given an FYI email may deal with a lot of this. And even if we did add this, it would have to be disabled by default, so the only way those 1500 cases would be tidied up is by contacting the people anyway.

**Comment by xsawyerx** [comment]
@ribasushi, I had meant lacking the ability of having "latest version" or "latest release" available for two versions. There can only be one "latest", even if you internally understand the idea of version branches ("1.0" vs. "2.0").

**Comment by karenetheridge** [comment]
One more important thing to remember: old tarballs might still contain a module that is mentioned in the 02packages index. More than one tarball for a distribution might contain a "latest version" of a module. It is important to not delete those, or the index will be affected.

For that matter, identifying these old modules that are still indexed and not overshadowed by a later release is something that would have value in its own right, as it often indicates an author error of some sort.

**Comment by ap** [comment]
@xsawyerx: and the reason for that is that the `perl` distribution is not installed through a CPAN client. It’s an odd duck in many ways, it’s distributed through the CPAN but not really part of it.
